stl/inc/scoped_allocator

```diff
@@ -250,16 +250,15 @@ template <class _Outer, class... _Inner>
 scoped_allocator_adaptor(_Outer, _Inner...) -> scoped_allocator_adaptor<_Outer, _Inner...>;
 #endif // _HAS_CXX17
 
-template <class _Outer1, class _Outer2, class _Inner1, class... _Inner>
-_NODISCARD bool operator==(const scoped_allocator_adaptor<_Outer1, _Inner1, _Inner...>& _Left,
-    const scoped_allocator_adaptor<_Outer2, _Inner1, _Inner...>& _Right) noexcept {
-    return _Left.outer_allocator() == _Right.outer_allocator() && _Left.inner_allocator() == _Right.inner_allocator();
-}
-
-template <class _Outer1, class _Outer2>
-_NODISCARD bool operator==(
-    const scoped_allocator_adaptor<_Outer1>& _Left, const scoped_allocator_adaptor<_Outer2>& _Right) noexcept {
-    return _Left.outer_allocator() == _Right.outer_allocator();
+template <class _Outer1, class _Outer2, class... _Inner>
+_NODISCARD bool operator==(const scoped_allocator_adaptor<_Outer1, _Inner...>& _Left,
+    const scoped_allocator_adaptor<_Outer2, _Inner...>& _Right) noexcept {
+    if constexpr (sizeof...(_Inner) == 0) {

```

Comment:
How is this ABI safe? 
I mean, the signature does not change, so old TU may be picked where this branch is absent.

Reply:
It is safe: exact specializations are instantiated in object files, not templates. No instantiation with zero-sized pack can be instantiated in an old TU.